File: src/main/java/no/novari/flyt/archive/gateway/template/validation/UniqueOrders.java
package no.novari.flyt.archive.gateway.template.validation;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.Retention;

import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Retention(RUNTIME)
@Constraint(validatedBy = UniqueOrdersValidator.class)
public @interface UniqueOrders {

    String DUPLICATE_ORDERS_REF = "duplicateOrders";

    String message() default "contains duplicate orders: {" + DUPLICATE_ORDERS_REF + "}";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};

}

File: src/main/kotlin/no/novari/flyt/archive/gateway/template/validation/UniqueOrders.kt
package no.novari.flyt.archive.gateway.template.validation

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

@Target(AnnotationTarget.CLASS)
@Retention(AnnotationRetention.RUNTIME)
@Constraint(validatedBy = [UniqueOrdersValidator::class])
annotation class UniqueOrders(
    val message: String = "contains duplicate orders: {$DUPLICATE_ORDERS_REF}",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = [],
) {
    companion object {
        const val DUPLICATE_ORDERS_REF = "duplicateOrders"
    }
}
